src/telempostdaemon.c

```diff
@@ -0,0 +1,632 @@
+        FILE *tmpfile = NULL;
+        int tmpfd = 0;
+
+        if (!daemon->is_spool_valid) {
+                /* If spool is not valid, simply drop the record */
+                return;
+        }
+
+        //check if the size is greater than 1 MB/spool max size
+        max_spool_size = spool_max_size_config();
+        if (max_spool_size != -1) {
+                telem_log(LOG_DEBUG, "Total size of spool dir: %ld\n", daemon->current_spool_size);
+                if (daemon->current_spool_size < 0) {
+                        telem_log(LOG_ERR, "Error getting spool directory size: %s\n",
+                                  strerror(-(int)(daemon->current_spool_size)));
+                        return;
+                } else if (daemon->current_spool_size >= (max_spool_size * 1024)) {
+                        telem_log(LOG_INFO, "Spool dir full, dropping record\n");
+                        return;
+                }
+        }
+
+        // create file with record
+        ret = asprintf(&tmpbuf, "%s/XXXXXX", spool_dir_config());
+        if (ret == -1) {
+                telem_log(LOG_ERR, "Failed to allocate memory for record name, aborting\n");
+                exit(EXIT_FAILURE);
+        }
+
+        tmpfd = mkstemp(tmpbuf);
+        if (tmpfd == -1) {
+                telem_perror("Error while creating temp file");
+                goto spool_err;
+        }
+
+        tmpfile = fdopen(tmpfd, "a");
+        if (!tmpfile) {
+                telem_perror("Error opening temp file");
+                close(tmpfd);
+                if (unlink(tmpbuf)) {
+                        telem_perror("Error deleting temp file");
+                }
+                goto spool_err;
+        }
+
+        // write headers
+        for (int i = 0; i < NUM_HEADERS; i++) {
+                fprintf(tmpfile, "%s\n", headers[i]);
+        }
+
+        //write body
+        fprintf(tmpfile, "%s\n", body);
+        fflush(tmpfile);
+        fclose(tmpfile);
+
+        /* The stat should not fail here unless it is ENOMEM */
+        if (stat(tmpbuf, &stat_buf) == 0) {
+                daemon->current_spool_size += (stat_buf.st_blocks * 512);
+        }
+
+spool_err:
+        free(tmpbuf);
+        return;
+}
+
+static void save_entry_to_journal(TelemPostDaemon *daemon, time_t t_stamp, char *headers[])
+{
+        char *classification_value;
+        char *event_id_value;
+
+        if (get_header_value(headers[TM_CLASSIFICATION], &classification_value) &&
+            get_header_value(headers[TM_EVENT_ID], &event_id_value)) {
+                if (new_journal_entry(daemon->record_journal, classification_value, t_stamp, event_id_value) != 0) {
+                        telem_log(LOG_INFO, "new_journal_entry in process_record: failed saving record entry\n");
+                }
+        }
+        free(classification_value);
+        free(event_id_value);
+}
+
+bool process_staged_record(char *filename, TelemPostDaemon *daemon)
+{
+        int k;
```

Comment:
I would prefer shorter functions to review.

Reply:
Agree, I refactored the function a little bit.